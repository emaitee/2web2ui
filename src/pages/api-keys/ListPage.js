import copy from 'copy-to-clipboard';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Banner, Page } from '@sparkpost/matchbox';

import { hideNewApiKey, listApiKeys } from 'src/actions/api-keys';

import ApiErrorBanner from 'src/components/apiErrorBanner/ApiErrorBanner';
import TableCollection from 'src/components/collection/TableCollection';
import Layout from 'src/components/layout/Layout';
import { getLoading } from 'src/selectors/api-keys';
import PermissionsColumn from './components/PermissionsColumn';

const columns = ['Name', 'Key', 'Permissions'];

const primaryAction = {
  content: 'Create API Key',
  Component: Link,
  to: '/account/api-keys/create'
};

const getRowData = (key) => [
  <Link to={`/account/api-keys/details/${key.id}`}>{key.label}</Link>,
  <code>{key.short_key} ••••</code>,
  <PermissionsColumn grants={key.grants} />
];

export class ListPage extends Component {
  state = { copied: false };

  componentDidMount() {
    this.props.listApiKeys();
  }

  onClickBanner = () => {
    copy(this.props.newKey);
    this.setState({ copied: true });
  };

  onReloadApiBanner = () => {
    this.props.listApiKeys({ force: true }); // force a refresh
  };

  renderBanner() {
    const { hideNewApiKey, newKey } = this.props;

    const action = {
      content: this.state.copied ? 'Copied to clipboard' : 'Copy',
      onClick: this.onClickBanner
    };

    return (
      <Banner
        action={action}
        title="New API Key"
        status="success"
        onDismiss={hideNewApiKey}
      >
        <p>
          Make sure to copy your API key now. You won't be able to see it again!
        </p>
        <strong>{newKey}</strong>
      </Banner>
    );
  }

  renderCollection() {
    const { keys } = this.props;

    return (
      <TableCollection
        columns={columns}
        getRowData={getRowData}
        pagination={true}
        rows={keys}
      />
    );
  }

  renderError() {
    return (
      <ApiErrorBanner
        errorDetails={this.props.error.message}
        message="Sorry, we seem to have had some trouble loading your API keys."
        reload={this.onReloadApiBanner}
      />
    );
  }

  render() {
    const { error, loading, newKey } = this.props;

    return (
      <Layout.App loading={loading}>
        <Page primaryAction={primaryAction} title="API Keys" />
        {newKey && this.renderBanner()}
        {error ? this.renderError() : this.renderCollection()}
      </Layout.App>
    );
  }
}

const mapStateToProps = (state) => {
  const { error, keys, newKey } = state.apiKeys;
  return {
    error,
    keys,
    loading: getLoading(state),
    newKey
  };
};

export default connect(mapStateToProps, { hideNewApiKey, listApiKeys })(
  ListPage
);
