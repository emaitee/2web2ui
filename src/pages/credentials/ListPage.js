import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Page } from '@sparkpost/matchbox';

import { fetchApiKeys } from 'actions/credentials';

import ApiErrorBanner from 'components/apiErrorBanner/ApiErrorBanner';
import TableCollection from 'components/collection/TableCollection';
import Layout from 'components/layout/Layout';
import PermissionsColumn from './components/PermissionsColumn';

const columns = ['Name', 'Key', 'Permissions'];

const primaryAction = {
  content: 'Create API Key',
  Component: Link,
  to: '/account/credentials/create'
};

const getRowData = (key) => [
  <Link to={`/account/credentials/details/${key.id}`}>{key.label}</Link>,
  <code>{key.short_key} ••••</code>,
  <PermissionsColumn grants={key.grants} />
];

export class ListPage extends Component {
  componentDidMount() {
    this.props.fetchApiKeys();
  }

  renderCollection() {
    const { keys } = this.props;

    return (
      <TableCollection
        columns={columns}
        getRowData={getRowData}
        pagination={true}
        rows={keys}
      />
    );
  }

  renderError() {
    const { error, fetchApiKeys } = this.props;

    return (
      <ApiErrorBanner
        errorDetails={error.message}
        message="Sorry, we seem to have had some trouble loading your API keys."
        reload={fetchApiKeys}
      />
    );
  }

  render() {
    const { error, loading } = this.props;

    return (
      <Layout.App loading={loading}>
        <Page primaryAction={primaryAction} title="API Keys" />
        {error ? this.renderError() : this.renderCollection()}
      </Layout.App>
    );
  }
}

const mapStateToProps = (state) => ({
  error: state.credentials.error,
  keys: state.credentials.keys,
  loading: state.credentials.loadingKeys
});

export default connect(mapStateToProps, { fetchApiKeys })(ListPage);
