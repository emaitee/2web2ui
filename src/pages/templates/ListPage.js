import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter, Link } from 'react-router-dom';
import config from 'src/config';

// Actions
import { listTemplates } from 'src/actions/templates';
import { getTemplates, getfilterTemplates } from 'src/selectors/templates';

// Components
import { Layout, TableCollection, ApiErrorBanner } from 'src/components';
import { Page, EmptyState } from '@sparkpost/matchbox';
import Filters from './components/Filters';
import { getRowData, TableHeader } from './components/TableComponents';

const primaryAction = {
  content: 'Create Template',
  to: '/templates/create',
  Component: Link
};

export class ListPage extends Component {

  componentDidMount() {
    this.props.listTemplates();
  }

  renderError() {
    return (
      <ApiErrorBanner
        message={'Sorry, we seem to have had some trouble loading your templates.'}
        errorDetails={this.props.error.message}
        reload={this.props.listTemplates}
      />
    );
  }

  renderCollection() {
    return (
      <TableCollection
        headerComponent={<TableHeader />}
        rows={this.props.templates}
        getRowData={getRowData}
        pagination
      />
    );
  }

  render() {
    const { count, loading, error } = this.props;

    if (!loading && count === 0) {
      return (
        <Layout.App>
          <EmptyState
            title='Manage your email templates'
            action={primaryAction} >
            {/* secondaryAction={content: 'Learn More', onClick: Learn_More()} > */}
            <p>Build, test, preview and send your transmissions.</p>
          </EmptyState>
        </Layout.App>
      );
    }

    return (
      <Layout.App loading={loading}>
        <Page
          primaryAction={primaryAction}
          title='Templates'
        />
        {error && this.renderError()}
        {!error && count > config.filters.minCount && <Filters />}
        {!error && this.renderCollection()}
      </Layout.App>
    );
  }
}

function mapStateToProps(state) {
  return {
    count: getTemplates(state).length,
    templates: getfilterTemplates(state),
    loading: state.templates.listLoading,
    error: state.templates.listError
  };
}

export default withRouter(connect(mapStateToProps, { listTemplates })(ListPage));
