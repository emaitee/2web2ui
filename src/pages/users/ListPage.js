import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import fp from 'lodash/fp';
import { Page } from '@sparkpost/matchbox';
import TimeAgo from 'react-timeago';

import * as usersActions from 'src/actions/users';
import { selectUsers } from 'src/selectors/users';

import { Loading, ApiErrorBanner, DeleteModal, TableCollection } from 'src/components';
import RoleSelect from './components/RoleSelect';
import DeleteButton from './components/DeleteButton';

const COLUMNS = ['Name', 'Role', 'Email', 'Last Login', null];
const DEFAULT_STATE = {
  userToDelete: {}
};

const primaryAction = {
  content: 'Invite User',
  Component: Link,
  to: '/account/users/create'
};

export class ListPage extends Component {
  state = DEFAULT_STATE;

  componentDidMount() {
    this.props.listUsers();
  }

  // Do not allow current user to change their access/role or delete their account
  getRowData = (user) => [
    user.name,
    <RoleSelect
      disabled={user.isCurrentUser}
      name={user.username}
      onChange={this.handleAccessChange}
      value={user.access}
      allowSuperUser={user.access === 'superuser'}
    />,
    user.email,
    user.last_login ? <TimeAgo date={user.last_login} live={false} /> : 'Never',
    <DeleteButton
      disabled={user.isCurrentUser}
      name={user.username}
      onClick={this.handleDeleteRequest}
    />
  ];

  handleAccessChange = ({ name, value }) => {
    this.props.updateUser(name, { access_level: value });
  }

  handleCancel = () => { this.setState(DEFAULT_STATE); }

  handleDelete = () => {
    const { userToDelete } = this.state;

    this.setState(DEFAULT_STATE, () => {
      this.props.deleteUser(userToDelete.username);
    });
  };

  handleDeleteRequest = (username) => {
    const user = fp.find((user) => user.username === username)(this.props.users);
    this.setState({ userToDelete: user });
  }

  renderError() {
    const { error, listUsers } = this.props;

    return (
      <ApiErrorBanner
        errorDetails={error.message}
        message="Sorry, we seem to have had some trouble loading your users."
        reload={listUsers}
      />
    );
  }

  // @note This component must always be the page to properly handle css transition
  renderDeleteModal() {
    const { userToDelete } = this.state;
    const isOpen = !fp.isEmpty(userToDelete);
    const name = fp.get('name')(userToDelete);

    return (
      <DeleteModal
        onDelete={this.handleDelete}
        onCancel={this.handleCancel}
        open={isOpen}
        content={<p>User "{name}" will no longer be able to log in or access this SparkPost account and all API keys associated with this user will be immediately deleted.</p>}
        title="Are you sure you want to delete this user?"
      />
    );
  }

  renderPage() {
    return (
      <div>
        <TableCollection
          columns={COLUMNS}
          getRowData={this.getRowData}
          pagination={true}
          rows={this.props.users}
          filterBox={{
            show: true,
            keyMap: { role: 'access' },
            exampleModifiers: ['name', 'email', 'role'],
            itemToStringKeys: ['username', 'name', 'email']
          }}
        />
        { this.renderDeleteModal() }
      </div>
    );
  }

  render() {
    const { loading, error, users } = this.props;

    if (loading) {
      return <Loading />;
    }

    return (
      <Page
        primaryAction={primaryAction}
        title="Users"
        empty={{
          show: users.length === 1,
          title: 'Invite Your Team to SparkPost',
          image: 'Users',
          content: <p>Manage your team's accounts and roles.</p>
        }}>
        { error ? this.renderError() : this.renderPage() }
      </Page>
    );
  }
}

const mapStateToProps = (state) => ({
  error: state.users.error,
  loading: state.users.loading,
  users: selectUsers(state)
});

export default connect(mapStateToProps, usersActions)(ListPage);
