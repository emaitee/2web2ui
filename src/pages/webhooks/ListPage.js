import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';

// Actions
import { listWebhooks } from '../../actions/webhooks';

// Components
import { Layout, TableCollection, ApiErrorBanner } from 'src/components';
import { Page } from '@sparkpost/matchbox';

const columns = ['Name', 'ID', 'Target'];
const getRowData = ({ id, name, target }) => {
  const nameLink = <Link to={`/webhooks/details/${id}`}>{name}</Link>;
  return [nameLink, id, target];
};

class WebhooksList extends Component {

  componentDidMount() {
    this.props.listWebhooks();
  }

  renderError() {
    const { error, listWebhooks } = this.props;
    return (
      <ApiErrorBanner
        message={'Sorry, we seem to have had some trouble loading your webhooks.'}
        errorDetails={error.message}
        reload={listWebhooks}
      />
    );
  }

  renderCollection() {
    const { webhooks } = this.props;
    return (
      <TableCollection
        columns={columns}
        rows={webhooks}
        getRowData={getRowData}
        pagination={true}
      />
    );
  }

  render() {
    const { webhooks, loading, error } = this.props;

    return (
      <Layout.App loading={webhooks.length === 0 && loading}>
        <Page
          primaryAction={{ content: 'Create Webhook', Component: Link, to: '/webhooks/create' }}
          title={'Webhooks'}
        />
        {error && this.renderError()}
        {!error && this.renderCollection()}
      </Layout.App>
    );
  }
}

function mapStateToProps({ webhooks }) {
  return {
    webhooks: webhooks.list,
    loading: webhooks.listLoading,
    error: webhooks.listError
  };
}

export default withRouter(connect(mapStateToProps, { listWebhooks })(WebhooksList));
