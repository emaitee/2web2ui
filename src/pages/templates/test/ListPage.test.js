import { shallow } from 'enzyme';
import React from 'react';

import { ListPage } from '../ListPage';

const props = {
  loading: false,
  error: null,
  count: 30,
  listTemplates: jest.fn(),
  templates: [
    {
      name: 'Temp 1',
    },
    {
      name: 'Temp 2',
    }
  ]
};

let wrapper;

beforeEach(() => {
  wrapper = shallow(<ListPage {...props} />);
});

it('renders correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

it('renders empty state', () => {
  wrapper.setProps({ count: 0 });
  expect(wrapper).toMatchSnapshot();
});

it('renders errors when present', () => {
  wrapper.setProps({ error: { message: 'Uh oh! It broke. ' }});
  expect(wrapper).toMatchSnapshot();
});
