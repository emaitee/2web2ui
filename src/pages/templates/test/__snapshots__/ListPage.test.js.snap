// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<App
  loading={false}
>
  <t
    primaryAction={
      Object {
        "Component": [Function],
        "content": "Create Template",
        "to": "/templates/create",
      }
    }
    title="Templates"
  />
  <Connect(ReduxForm) />
  <TableCollection
    getRowData={[Function]}
    headerComponent={<TableHeader />}
    pagination={true}
    rows={
      Array [
        Object {
          "name": "Temp 1",
        },
        Object {
          "name": "Temp 2",
        },
      ]
    }
  />
</App>
`;

exports[`renders empty state 1`] = `
<App>
  <t
    action={
      Object {
        "Component": [Function],
        "content": "Create Template",
        "to": "/templates/create",
      }
    }
    title="Manage your email templates"
  >
    <p>
      Build, test, preview and send your transmissions.
    </p>
  </t>
</App>
`;

exports[`renders errors when present 1`] = `
<App
  loading={false}
>
  <t
    primaryAction={
      Object {
        "Component": [Function],
        "content": "Create Template",
        "to": "/templates/create",
      }
    }
    title="Templates"
  />
  <ApiErrorBanner
    errorDetails="Uh oh! It broke. "
    message="Sorry, we seem to have had some trouble loading your templates."
    reload={[Function]}
  />
</App>
`;
